include/deal.II/multigrid/mg_coarse.h

```diff
@@ -356,6 +356,58 @@ MGCoarseGridIterativeSolver<VectorType,
 
 
 
+namespace internal
+{
+  namespace MGCoarseGridIterativeSolver
+  {
+    template <
+      class VectorType,
+      class SolverType,
+      class MatrixType,
+      class PreconditionerType,
+      typename std::enable_if<
+        std::is_same<VectorType, typename SolverType::vector_type>::value,
+        VectorType>::type * = nullptr>
+    void
+    solve(SolverType &              solver,
+          const MatrixType &        matrix,
+          const PreconditionerType &preconditioner,
+          VectorType &              dst,
+          const VectorType &        src)
```

Comment:
DId you try to add the specialization of the form
```suggestion
      typename std::enable_if<
        std::is_same<VectorType, typename SolverType::vector_type>::value,
        VectorType>::type * = nullptr>
    void
    solve(SolverType &              solver,
          const MatrixType &        matrix,
          const PreconditionerType &preconditioner,
          typename SolverType::vector_type &              dst,
          const typename SolverType::vector_type &        src)
```
That way, we could skip the `std::enable_if` part and make it look slightly more intuitive. But I'm not sure if the compiler has enough information to distinguish the two cases and pick the more specialized one.

Reply:
I have not tried in this instance. But in other instances (like when working on `ReadWrtiteVector`, `DataOut`, ...), I have seen that the compiler has hard time to decide in such cases, since it could pick any of the both functions.